File: app-ios-nativa/Dominio/Pruebas/PruebasDominio/PruebasPresentacion.swift
import Foundation
import Testing

@testable import Dominio

/// Mediodía UTC del 15 de marzo de 2024. A mediodía, y con la zona horaria
/// fijada en la prueba, la fecha no se va al día de antes ni al de después
/// según dónde se ejecute esto.
private let mediodiaUtc: MarcaDeTiempo = 1_710_504_000_000

private func elemento(
    url: String,
    titulo: String? = nil,
    etiquetas: [String] = [],
    creadoEn: MarcaDeTiempo = mediodiaUtc
) -> Elemento {
    Elemento(
        id: "e1",
        url: url,
        titulo: titulo,
        etiquetas: etiquetas,
        creadoEn: creadoEn,
        // Muy posterior a la de guardado: la fila no debe enseñar esta.
        actualizadoEn: creadoEn + 30 * 86_400_000
    )
}

private func subtitulo(_ elemento: Elemento) -> String {
    Presentacion.subtitulo(de: elemento, zonaHoraria: TimeZone(identifier: "UTC")!)
}

@Suite("Título de la fila")
struct PruebasTituloFila {
    @Test("usa el título cuando lo hay")
    func conTitulo() {
        #expect(Presentacion.titulo(de: elemento(url: "https://a.com", titulo: "A")) == "A")
    }

    @Test("sin título usa la URL, que es mejor que una fila muda")
    func sinTitulo() {
        #expect(Presentacion.titulo(de: elemento(url: "https://a.com")) == "https://a.com")
    }

    @Test("un título vacío cuenta como no tener título")
    func tituloVacio() {
        let sinNada = elemento(url: "https://a.com", titulo: "")

        #expect(Presentacion.titulo(de: sinNada) == "https://a.com")
    }
}

@Suite("Subtítulo de la fila")
struct PruebasSubtituloFila {
    @Test("lleva dominio, etiquetas y fecha, en ese orden")
    func completo() {
        let uno = elemento(
            url: "https://www.ejemplo.com/articulo",
            titulo: "Un artículo interesante",
            etiquetas: ["ocio", "pendiente"]
        )

        #expect(subtitulo(uno) == "www.ejemplo.com — ocio, pendiente — 15 de marzo de 2024")
    }

    @Test("la fecha lleva el mes en letra, que es como se oye bien")
    func mesEnLetra() {
        #expect(subtitulo(elemento(url: "https://a.com")).contains("de marzo de"))
    }

    @Test("sin título el subtítulo sigue diciendo el dominio")
    func sinTitulo() {
        #expect(subtitulo(elemento(url: "https://a.com")).hasPrefix("a.com"))
    }

    @Test("sin etiquetas no queda un separador vacío en medio")
    func sinEtiquetas() {
        #expect(!subtitulo(elemento(url: "https://a.com", titulo: "A")).contains(" —  — "))
    }

    @Test("sin fecha lo dice, en vez de soltar 1970")
    func sinFecha() {
        #expect(subtitulo(elemento(url: "https://a.com", creadoEn: 0)).contains("sin fecha"))
    }

    @Test("la fecha es la de cuando se guardó, no la del último cambio")
    func fechaDeGuardadoNoDeCambio() {
        // Cambiarle una etiqueta a un enlace de hace un mes movía la fecha que
        // se lee en la fila, y entonces ya no había forma de saber cuándo se
        // había guardado de verdad.
        let guardadoEnMarzo = elemento(url: "https://a.com", titulo: "A")
        let retocadoHoy = guardadoEnMarzo.conEtiquetas(
            ["ocio"], ahora: { mediodiaUtc + 99_999_999 })

        #expect(subtitulo(retocadoHoy).contains("15 de marzo de 2024"))
    }

    @Test("lo que no es una URL no aporta dominio, pero tampoco rompe la fila")
    func urlIlegible() {
        let raro = elemento(url: "esto no es una url", titulo: "Raro")

        #expect(subtitulo(raro) == "15 de marzo de 2024")
    }
}

@Suite("Textos del detalle")
struct PruebasTextosDetalle {
    @Test("la fecha de guardado se lee, no se descifra")
    func fechaDeGuardado() {
        let fecha = Presentacion.fechaLegible(
            mediodiaUtc,
            locale: Presentacion.localeDeLaApp,
            zonaHoraria: TimeZone(identifier: "UTC")!
        )

        #expect(Textos.guardadoEl(fecha) == "Guardado el 15 de marzo de 2024")
    }

    @Test("un enlace sin fecha lo dice en vez de soltar 1970")
    func sinFecha() {
        let fecha = Presentacion.fechaLegible(
            0,
            locale: Presentacion.localeDeLaApp,
            zonaHoraria: TimeZone(identifier: "UTC")!
        )

        #expect(fecha == "sin fecha")
    }
}

File: app-android/dominio/src/test/kotlin/com/jmortizsilva/guardarenlaces/dominio/PruebasPresentacion.kt
package com.jmortizsilva.guardarenlaces.dominio

import java.time.ZoneOffset
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Mediodía UTC del 15 de marzo de 2024. A mediodía, y con la zona fijada en la prueba, la fecha no
 * se va al día de antes ni al de después según dónde se ejecute esto.
 */
private const val MEDIODIA_UTC: MarcaDeTiempo = 1_710_504_000_000

private fun elemento(
    url: String,
    titulo: String? = null,
    etiquetas: List<String> = emptyList(),
    creadoEn: MarcaDeTiempo = MEDIODIA_UTC,
) =
    Elemento(
        id = "e1",
        url = url,
        titulo = titulo,
        etiquetas = etiquetas,
        creadoEn = creadoEn,
        // Muy posterior a la de guardado: la fila no debe enseñar esta.
        actualizadoEn = creadoEn + 30 * 86_400_000L,
    )

private fun subtitulo(elemento: Elemento) = Presentacion.subtitulo(elemento, zona = ZoneOffset.UTC)

class PruebasTituloFila {
    @Test
    fun `usa el titulo cuando lo hay`() {
        assertEquals("A", Presentacion.titulo(elemento("https://a.com", titulo = "A")))
    }

    @Test
    fun `sin titulo usa la URL, que es mejor que una fila muda`() {
        assertEquals("https://a.com", Presentacion.titulo(elemento("https://a.com")))
    }

    @Test
    fun `un titulo vacio cuenta como no tener titulo`() {
        assertEquals("https://a.com", Presentacion.titulo(elemento("https://a.com", titulo = "")))
    }
}

class PruebasSubtituloFila {
    @Test
    fun `lleva dominio, etiquetas y fecha, en ese orden`() {
        val uno =
            elemento(
                "https://www.ejemplo.com/articulo",
                titulo = "Un artículo interesante",
                etiquetas = listOf("ocio", "pendiente"),
            )

        assertEquals("www.ejemplo.com — ocio, pendiente — 15 de marzo de 2024", subtitulo(uno))
    }

    @Test
    fun `la fecha lleva el mes en letra, que es como se oye bien`() {
        assertTrue("de marzo de" in subtitulo(elemento("https://a.com")))
    }

    @Test
    fun `sin titulo el subtitulo sigue diciendo el dominio`() {
        assertTrue(subtitulo(elemento("https://a.com")).startsWith("a.com"))
    }

    @Test
    fun `sin etiquetas no queda un separador vacio en medio`() {
        assertFalse(" —  — " in subtitulo(elemento("https://a.com", titulo = "A")))
    }

    @Test
    fun `sin fecha lo dice, en vez de soltar 1970`() {
        assertTrue("sin fecha" in subtitulo(elemento("https://a.com", creadoEn = 0)))
    }

    @Test
    fun `la fecha es la de cuando se guardo, no la del ultimo cambio`() {
        // Cambiarle una etiqueta a un enlace de hace un mes movía la fecha que se lee en la fila, y
        // entonces ya no había forma de saber cuándo se había guardado de verdad.
        val retocadoHoy =
            elemento("https://a.com", titulo = "A")
                .conEtiquetas(listOf("ocio"), ahora = { MEDIODIA_UTC + 99_999_999 })

        assertTrue("15 de marzo de 2024" in subtitulo(retocadoHoy))
    }

    @Test
    fun `lo que no es una URL no aporta dominio, pero tampoco rompe la fila`() {
        assertEquals(
            "15 de marzo de 2024",
            subtitulo(elemento("esto no es una url", titulo = "Raro")),
        )
    }

    @Test
    fun `la fecha va en español aunque el telefono este en otro idioma`() {
        // La zona y el idioma del teléfono no se usan para la fecha: se fija el español de la app.
        assertEquals("es", Presentacion.localeDeLaApp.language)
        assertEquals("ES", Presentacion.localeDeLaApp.country)
    }
}

class PruebasTextosDetalle {
    @Test
    fun `la fecha de guardado se lee, no se descifra`() {
        val fecha =
            Presentacion.fechaLegible(MEDIODIA_UTC, Presentacion.localeDeLaApp, ZoneOffset.UTC)

        assertEquals("Guardado el 15 de marzo de 2024", Textos.guardadoEl(fecha))
    }

    @Test
    fun `un enlace sin fecha lo dice en vez de soltar 1970`() {
        assertEquals(
            "sin fecha",
            Presentacion.fechaLegible(0, Presentacion.localeDeLaApp, ZoneOffset.UTC),
        )
    }
}
